kgforge/specializations/models/rdf/service.py

```diff
@@ -272,3 +343,121 @@ def traverse_properties(properties) -> Tuple[Dict, Dict]:
+                ) from ke
+            except ParserError as pe:
+                raise ValueError(
+                    f"Failed to parse the rdf graph of the imported schema {imported_schema_uriref}: {str(pe)}"
+                ) from pe
+        if node_shape_uriref:
+            triples_to_add, _, triples_to_remove = (
+                self._get_property_shapes_from_nodeshape(node_shape_uriref)
+            )
+            for triple_to_add in triples_to_add:
+                schema_graph.add(triple_to_add)
+            for triple_to_remove in triples_to_remove:
+                schema_graph.remove(triple_to_remove)
+        return schema_graph
+
+    def _get_property_shapes_from_nodeshape(self, node_shape_uriref):
+        property_shapes_to_add = []
+        triples_to_add = []
+        triples_to_remove = []
+        schema_defines_shape = [
+            SH["and"] / (RDF.first | RDF.rest / RDF.first) * ZeroOrMore,
+            SH["or"] / (RDF.first | RDF.rest / RDF.first) * ZeroOrMore,
+            SH["xone"] / (RDF.first | RDF.rest / RDF.first) * ZeroOrMore,
+        ]
+        sh_nodes = []
+        sh_properties = []
+        sh_properties.extend(list(self._graph.objects(node_shape_uriref, SH.property)))
+        sh_nodes.extend(list(self._graph.objects(node_shape_uriref, SH.node)))
+        for s in schema_defines_shape:
+            sh_nodes.extend(list(self._graph.objects(node_shape_uriref, s / SH.node)))
+            sh_properties.extend(
+                list(self._graph.objects(node_shape_uriref, s / SH.property))
+            )
+        for sh_node in set(sh_nodes):
+            if str(sh_node) != str(node_shape_uriref) and str(sh_node) != str(RDF.nil):
+                t_a, p_a, t_r = self._get_property_shapes_from_nodeshape(sh_node)
+                if p_a:
+                    triples_to_add.extend(t_a)
+                    for propertyShape in p_a:
+                        triples_to_add.append(
+                            (node_shape_uriref, propertyShape[0], propertyShape[1])
+                        )
+                        property_shapes_to_add.append(propertyShape)
+                    triples_to_remove.extend(t_r)
+                    triples_to_remove.append((node_shape_uriref, SH.node, sh_node))
+
+        for sh_node_property in set(sh_properties):
+            if str(sh_node_property) != str(node_shape_uriref) and str(
+                sh_node_property
+            ) != str(RDF.nil):
+                property_shapes_to_add.append((SH.property, sh_node_property))
+
+        return triples_to_add, property_shapes_to_add, triples_to_remove
+
+    def get_shape_graph(self, shape_uriref: URIRef) -> Tuple[Shape, Graph]:
+        try:
+            shape = self.get_shape_graph_wrapper().lookup_shape_from_node(shape_uriref)
+            if (
+                shape_uriref in self.shape_to_defining_resource
+                and self.shape_to_defining_resource[shape_uriref] in self._imported
```

Comment:
I wonder about the reason for the existence of the variable self._imported.
I would think that checking the states of the dictionaries or the graph dataset could let us know about whether something is imported, but the maintenance of this extra array may lead to things not being in sync (is it a valid state to have an entry in self.shape_to_defining_resource, but not to be in self._imported, or the other way around?)

Reply:
I see that self._imported.append() only happens after `self.load_shape_graph` and inside it creates one graph into `_graph`. So maybe there can be a method called is_imported that checks whether there is an entry inside `self._graph` for a named graph using for `defining_resource_to_named_graph[schema_uriref]`

Reply:
Then it depends what the definition of imported is, does that also include transitive imports?

Reply:
I moved it to the end of the method